commit d6f64b89d7ff22ce05896ab4a93a653e8d0b123d
Author: Al Viro <viro@zeniv.linux.org.uk>
Date:   Tue Feb 15 22:26:01 2011 -0500

    close race in /proc/*/environ
    
    Switch to mm_for_maps().  Maybe we ought to make it r--r--r--,
    since we do checks on IO anyway...
    
    Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
    [dannf: backported to Debian's 2.6.26]

commit c6e0832fdcd651328728c00e6464f36c091444fa
Author: dann frazier <dannf@debian.org>
Date:   Wed Aug 31 22:10:48 2011 -0600

    bugfix/all/close-race-in-proc-pid-environ.patch

diff --git a/fs/proc/base.c b/fs/proc/base.c
index 6e71515..f582fff 100644
--- a/fs/proc/base.c
+++ b/fs/proc/base.c
@@ -906,20 +906,18 @@ static ssize_t environ_read(struct file *file, char __user *buf,
 	if (!task)
 		goto out_no_task;
 
-	if (!ptrace_may_attach(task))
-		goto out;
-
 	ret = -ENOMEM;
 	page = (char *)__get_free_page(GFP_TEMPORARY);
 	if (!page)
 		goto out;
 
-	ret = 0;
 
-	mm = get_task_mm(task);
-	if (!mm)
+	mm = mm_for_maps(task);
+	ret = PTR_ERR(mm);
+	if (!mm || IS_ERR(mm))
 		goto out_free;
 
+	ret = 0;
 	while (count > 0) {
 		int this_len, retval, max_len;
 
